docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+
+### Requirements
+
+#### Functional
+
+<a name="FR1">FR1.</a> CAPZ MUST support deleting an individual Virtual Machine Scale Set instance.
+
+<a name="FR2">FR2.</a> CAPZ MUST support cordon and draining workload from a Virtual Machine Scale Set instance.
+
+#### Non-Functional
+
+<a name="NFR1">NFR1.</a> CAPZ SHOULD provide resource status updates as the Azure resources are provisioned
+
+<a name="NFR2">NFR2.</a> CAPZ SHOULD not overwhelm Azure API request limits and should rate limit reconciliation cycles
+
+<a name="NFR3">NFR3.</a> Unit tests MUST exist upgrade and delete instance selection
+
+<a name="NFR4">NFR4.</a> e2e tests MUST exist MachinePool upgrade, scale up / down, and instance delete scenarios
+
+### Implementation Details/Notes/Constraints
+
+The current implementation of CAPZ AzureMachinePool embeds the state of each of the machines in the MachinePool within 
+the status of the AzureMachinePool.
+
+```go
+// AzureMachinePoolStatus defines the observed state of AzureMachinePool
+AzureMachinePoolStatus struct {
+	
+    /*
+        Other fields omitted for brevity    
+    */
+    
+    // Instances is the VM instance status for each VM in the VMSS
+    // +optional
+    Instances []*AzureMachinePoolInstanceStatus `json:"instances,omitempty"`
+}
+
+// AzureMachinePoolInstanceStatus provides status information for each instance in the VMSS
+AzureMachinePoolInstanceStatus struct {
+    // Version defines the Kubernetes version for the VM Instance
+    // +optional
+    Version string `json:"version"`
+    
+    // ProvisioningState is the provisioning state of the Azure virtual machine instance.
+    // +optional
+    ProvisioningState *infrav1.VMState `json:"provisioningState"`
+    
+    // ProviderID is the provider identification of the VMSS Instance
+    // +optional
+    ProviderID string `json:"providerID"`
+    
+    // InstanceID is the identification of the Machine Instance within the VMSS
+    // +optional
+    InstanceID string `json:"instanceID"`
+    
+    // InstanceName is the name of the Machine Instance within the VMSS
+    // +optional
+    InstanceName string `json:"instanceName"`
+    
+    // LatestModelApplied indicates the instance is running the most up-to-date VMSS model. A VMSS model describes
+    // the image version the VM is running. If the instance is not running the latest model, it means the instance
+    // may not be running the version of Kubernetes the Machine Pool has specified and needs to be updated.
+    LatestModelApplied bool `json:"latestModelApplied"`
+}
+```
+
+#### Proposed API Changes
+The proposed changes below show the AzureMachinePool and AzureMachinePoolMachine.
+
+```go
+// AzureMachinePoolSpec defines the desired state of AzureMachinePool
+AzureMachinePoolSpec struct {
+
+    /*
+       Other fields omitted for brevity
+    */
+	
+    // MaxUnavailable is the max number of replicas which can be unavailable at any time. For example, if the
```

Comment:
My immediate reaction when seeing these new types is "why put them in CAPZ? there is nothing Azure infra specific about it". I know there are nuances and adding these types directly in CAPI is not straight forward because it then implies different providers would need to honor these values by either a) passing them through to the infra properties if the Cloud/environment supports them or b) implement it themselves (what we're doing in CAPZ). 

I worry that adding these here now will make it hard to move them later since any refactor of the types to CAPI types would be either a breaking change. And the alternative of leaving them in here beyond exp/ is that we create diverging implementations and, more importantly, inconsistent user experience for MachinePool across providers.

If we do go down this path, I suggest detailing this in a risks & mitigations section in this doc so that we can justify this design choice down the road when someone inevitably asks "why was this not added to CAPI types?"

Reply:
Great observation. I totally agree. It'd be great to remove these from the AMP spec and move them to MP.

With https://github.com/kubernetes-sigs/cluster-api/pull/4073 nearly merged, I'll open up a PR to add a `RolloutStrategy` and `DeleteStrategy` to `MachinePool`.

Reply:
How is that action plan tracked in the proposal?

Reply:
It looks like [`AWSMachinePool` in CAPA contains similar fields](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/17c3144c5f0060526a005eb5a2927db22fcd259a/exp/api/v1alpha3/awsmachinepool_types.go#L82-L99). Not sure if it will be possible to move them to `MachinePool` as different cloud providers use very different mechanisms (ASG vs VMSS).

Reply:
That's a good point @fiunchinho, it might make sense to keep them in cloud providers then. Let's document this as an intentional choice in this proposal.